File: src/algorithms/graph/eulerian/hierholzers/__tests__/hierholzers_test.go
package hierholzers

import "testing"

func isValidCircuit(circuit []string, adjacencyList map[string][]string, startNodeId string) bool {
	if len(circuit) == 0 {
		return false
	}
	if circuit[0] != startNodeId {
		return false
	}
	if circuit[len(circuit)-1] != startNodeId {
		return false
	}
	totalEdges := 0
	for _, neighbors := range adjacencyList {
		totalEdges += len(neighbors)
	}
	totalEdges /= 2
	return len(circuit)-1 == totalEdges
}

func TestFindsEulerianCircuitOnSimpleTriangle(t *testing.T) {
	adj := map[string][]string{"A": {"B", "C"}, "B": {"A", "C"}, "C": {"B", "A"}}
	circuit := hierholzersAlgorithm(adj, "A")
	if circuit[0] != "A" || circuit[len(circuit)-1] != "A" {
		t.Errorf("Circuit must start and end at A, got %v", circuit)
	}
	if !isValidCircuit(circuit, adj, "A") {
		t.Error("Circuit is not valid Eulerian")
	}
}

func TestFindsEulerianCircuitOnDefault5NodeGraph(t *testing.T) {
	adj := map[string][]string{
		"A": {"B", "C", "D", "E"},
		"B": {"A", "C"},
		"C": {"B", "A"},
		"D": {"A", "E"},
		"E": {"D", "A"},
	}
	circuit := hierholzersAlgorithm(adj, "A")
	if circuit[0] != "A" || circuit[len(circuit)-1] != "A" {
		t.Error("Circuit must start and end at A")
	}
	if !isValidCircuit(circuit, adj, "A") {
		t.Error("Circuit is not valid Eulerian")
	}
}

func TestReturnsSingleNodeCircuitForGraphWithNoEdges(t *testing.T) {
	adj := map[string][]string{"A": {}}
	circuit := hierholzersAlgorithm(adj, "A")
	if len(circuit) != 1 || circuit[0] != "A" {
		t.Errorf("Expected [A], got %v", circuit)
	}
}

func TestFindsEulerianCircuitOnSquare(t *testing.T) {
	adj := map[string][]string{
		"A": {"B", "D"}, "B": {"A", "C"}, "C": {"B", "D"}, "D": {"C", "A"},
	}
	circuit := hierholzersAlgorithm(adj, "A")
	if circuit[0] != "A" || circuit[len(circuit)-1] != "A" {
		t.Error("Circuit must start and end at A")
	}
	if !isValidCircuit(circuit, adj, "A") {
		t.Error("Circuit is not valid Eulerian")
	}
}

func TestFindsEulerianCircuitOnTwoTrianglesSharingANode(t *testing.T) {
	adj := map[string][]string{
		"A": {"B", "C", "D", "E"},
		"B": {"A", "C"},
		"C": {"B", "A"},
		"D": {"A", "E"},
		"E": {"D", "A"},
	}
	circuit := hierholzersAlgorithm(adj, "A")
	if circuit[0] != "A" || circuit[len(circuit)-1] != "A" {
		t.Error("Circuit must start and end at A")
	}
	if len(circuit) != 7 {
		t.Errorf("Expected length 7, got %d", len(circuit))
	}
}

func TestFindsEulerianCircuitStartingFromNonHubNode(t *testing.T) {
	adj := map[string][]string{"A": {"B", "C"}, "B": {"A", "C"}, "C": {"B", "A"}}
	circuit := hierholzersAlgorithm(adj, "B")
	if circuit[0] != "B" || circuit[len(circuit)-1] != "B" {
		t.Error("Circuit must start and end at B")
	}
	if len(circuit) != 4 {
		t.Errorf("Expected length 4, got %d", len(circuit))
	}
}

func TestProducesCircuitOnlyIncludingNodesWithEdges(t *testing.T) {
	adj := map[string][]string{"A": {"B", "C"}, "B": {"A", "C"}, "C": {"B", "A"}}
	circuit := hierholzersAlgorithm(adj, "A")
	validNodes := map[string]bool{"A": true, "B": true, "C": true}
	for _, nodeId := range circuit {
		if !validNodes[nodeId] {
			t.Errorf("Unexpected node %s in circuit", nodeId)
		}
	}
}

File: src/algorithms/graph/eulerian/hierholzers/hierholzers_test.rs
include!("sources/hierholzers.rs");

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashMap;

    fn make_adj(pairs: &[(&str, &[&str])]) -> HashMap<String, Vec<String>> {
        pairs
            .iter()
            .map(|(node, neighbors)| {
                (node.to_string(), neighbors.iter().map(|n| n.to_string()).collect())
            })
            .collect()
    }

    fn is_valid_circuit(circuit: &[String], adjacency_list: &HashMap<String, Vec<String>>, start: &str) -> bool {
        if circuit.is_empty() { return false; }
        if circuit[0] != start { return false; }
        if circuit[circuit.len() - 1] != start { return false; }
        let total_edges: usize = adjacency_list.values().map(|v| v.len()).sum::<usize>() / 2;
        circuit.len() - 1 == total_edges
    }

    #[test]
    fn finds_eulerian_circuit_on_simple_triangle() {
        let adj = make_adj(&[("A", &["B", "C"]), ("B", &["A", "C"]), ("C", &["B", "A"])]);
        let circuit = hierholzers_algorithm(&adj, "A");
        assert_eq!(circuit[0], "A");
        assert_eq!(circuit[circuit.len() - 1], "A");
        assert!(is_valid_circuit(&circuit, &adj, "A"));
    }

    #[test]
    fn finds_eulerian_circuit_on_default_5_node_graph() {
        let adj = make_adj(&[
            ("A", &["B", "C", "D", "E"]),
            ("B", &["A", "C"]),
            ("C", &["B", "A"]),
            ("D", &["A", "E"]),
            ("E", &["D", "A"]),
        ]);
        let circuit = hierholzers_algorithm(&adj, "A");
        assert_eq!(circuit[0], "A");
        assert_eq!(circuit[circuit.len() - 1], "A");
        assert!(is_valid_circuit(&circuit, &adj, "A"));
    }

    #[test]
    fn returns_single_node_circuit_for_graph_with_no_edges() {
        let adj = make_adj(&[("A", &[])]);
        let circuit = hierholzers_algorithm(&adj, "A");
        assert_eq!(circuit, vec!["A".to_string()]);
    }

    #[test]
    fn finds_eulerian_circuit_on_square() {
        let adj = make_adj(&[
            ("A", &["B", "D"]), ("B", &["A", "C"]), ("C", &["B", "D"]), ("D", &["C", "A"]),
        ]);
        let circuit = hierholzers_algorithm(&adj, "A");
        assert_eq!(circuit[0], "A");
        assert_eq!(circuit[circuit.len() - 1], "A");
        assert!(is_valid_circuit(&circuit, &adj, "A"));
    }

    #[test]
    fn finds_eulerian_circuit_on_two_triangles_sharing_a_node() {
        let adj = make_adj(&[
            ("A", &["B", "C", "D", "E"]),
            ("B", &["A", "C"]),
            ("C", &["B", "A"]),
            ("D", &["A", "E"]),
            ("E", &["D", "A"]),
        ]);
        let circuit = hierholzers_algorithm(&adj, "A");
        assert_eq!(circuit[0], "A");
        assert_eq!(circuit[circuit.len() - 1], "A");
        assert_eq!(circuit.len(), 7);
    }

    #[test]
    fn finds_eulerian_circuit_starting_from_non_hub_node() {
        let adj = make_adj(&[("A", &["B", "C"]), ("B", &["A", "C"]), ("C", &["B", "A"])]);
        let circuit = hierholzers_algorithm(&adj, "B");
        assert_eq!(circuit[0], "B");
        assert_eq!(circuit[circuit.len() - 1], "B");
        assert_eq!(circuit.len(), 4);
    }

    #[test]
    fn produces_circuit_only_including_nodes_with_edges() {
        let adj = make_adj(&[("A", &["B", "C"]), ("B", &["A", "C"]), ("C", &["B", "A"])]);
        let circuit = hierholzers_algorithm(&adj, "A");
        let valid_nodes: std::collections::HashSet<&str> = ["A", "B", "C"].iter().copied().collect();
        for node_id in &circuit {
            assert!(valid_nodes.contains(node_id.as_str()));
        }
    }
}
